File: app/src/main/java/com/github/premnirmal/ticker/network/GStockConverter.java
package com.github.premnirmal.ticker.network;

import com.google.gson.reflect.TypeToken;

import java.io.IOException;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

import retrofit.converter.ConversionException;
import retrofit.mime.TypedInput;

/**
 * Created by premnirmal on 4/10/15.
 */
class GStockConverter extends BaseConverter {

    @Override
    public Object fromBody(TypedInput body, Type type) throws ConversionException {
        try {
            final String bodyString = getString(body.in()).replaceAll("\n","");
            final String responseString;
            if (bodyString.startsWith("//")) {
                responseString = bodyString.substring(2, bodyString.length());
            } else {
                responseString = bodyString;
            }
            final Type collectionType = new TypeToken<List<GStock>>() {
            }.getType();
            final List<GStock> stocks = gson.fromJson(responseString, collectionType);
            return stocks;
        } catch (IOException e) {
            e.printStackTrace();
            return new ArrayList<GStock>();
        }
    }

}

File: app/src/main/kotlin/com/github/premnirmal/ticker/network/GStockConverter.kt
package com.github.premnirmal.ticker.network

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import okhttp3.ResponseBody
import java.io.IOException
import java.lang.reflect.Type
import java.util.ArrayList

/**
 * Created on 3/3/16.
 */
internal class GStockConverter(gson: Gson) : BaseConverter<List<GStock>>(gson) {

  override fun convert(value: ResponseBody?): List<GStock> {
    try {
      val bodyString = getString(value!!.byteStream()).replace("\n".toRegex(), "")
      val responseString: String
      if (bodyString.startsWith("//")) {
        responseString = bodyString.substring(bodyString.indexOf('['),
            bodyString.lastIndexOf(']') + 1)
      } else {
        responseString = bodyString
      }
      val type = object : TypeToken<ArrayList<GStock>>() {}.type
      val stocks = gson.fromJson<List<GStock>>(responseString, type)
      return stocks
    } catch (e: IOException) {
      e.printStackTrace()
      throw error(e)
    }
  }
}